scripts/run_action.mjs

```diff
@@ -0,0 +1,90 @@
+// Copyright 2022 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import url from "url";
+
+const resolveActionPath = async actionPath => {
+  let result = "";
+
+  if (!actionPath) return result;
+
+  const currentDirectory = (await $`pwd`).toString().trim();
+  const roots = [currentDirectory, `${currentDirectory}/src`];
+
+  while (roots.length && !result) {
+    const root = roots.shift();
+    const pathCandidate = path.resolve(root, actionPath);
+    const extensions = ["sh", "mjs"];
+
+    while (extensions.length && !result) {
+      const extension = extensions.shift();
+
+      if (await fs.pathExists(`${pathCandidate}.action.${extension}`)) {
+        result = `${pathCandidate}.action.${extension}`;
+      }
+    }
+  }
+
+  return result;
+};
+
+/**
+ *
+ * @param {*} actionPath
+ * @returns
+ */
+export async function runAction(actionPath) {
+  const resolvedPath = await resolveActionPath(actionPath);
+  if (!resolvedPath) {
+    console.info("Please provide an action to run.");
+    return runAction("list");
+  }
+
+  const startTime = performance.now();
+  console.group(chalk.yellow.bold(`▶ action(${actionPath}):`));
+
+  try {
+    if (resolvedPath.endsWith("mjs")) {
+      const module = await import(resolvedPath);
+
+      await module.main();
+    } else {
+      await $`bash ${resolvedPath} ${argv._.slice(2)}`;
```

Comment:
This is the one and only place where we need zx.
I'd prefer to avoid zx here. It's harder to reason about argument quoting and people are not familiar with it. It's probably easier to misuse. How would it look like with the standard functions?

Reply:
zx takes care of quoting and escaping for us, so without that we'd need to build that logic ourselves. https://github.com/google/zx/blob/main/docs/quotes.md

But apart from that, i think it's just `child_process.spawn()`

Reply:
Oh, and we'd have to promisify it ourselves too, I think.

Reply:
Please remove zx. Too magical and not providing much value.

Use the better-understood `child_process.spawnSync()` instead (https://nodejs.org/api/child_process.html#child_processspawnsynccommand-args-options).
There's no need to quote because you can pass a list of args. It's much better than crafting command lines.
I guess you can use `options.shell = true`

Reply:
Okie doke, should be pretty easy.

Reply:
I got caught up in the idea of using our own other open source stuff, I guess

Reply:
done

Reply:
So I opted to promisify spawn because spawnSync returns stdout all at once, while with spawn i can log to the console as data comes in from the child process.

Reply:
Pass `options.stdio = "inherit"` and I believe it will output as needed:
https://nodejs.org/api/child_process.html#optionsstdio

Then you don't need the spawnSync and the stream function you created.

Reply:
Hm, it didn't work it seems
<img width="953" alt="Screen Shot 2022-03-28 at 11 05 19" src="https://user-images.githubusercontent.com/3759828/160428353-a55b6521-7483-418a-8f70-e8df1e5508b3.png">
:

Reply:
It seems like it was output, but outside the console group?
Is the console group not output as it happens?

Reply:
Ah, derp, I had the wrong option, should have been `inherit`. Sorry - operating on low sleep today.

So that works and preserves color (!) -  however, it breaks the indenting, since technically it is logging at the process level rather than within the console group.

I guess some sort of `process.env.INDENT_LEVEL++` would fix that.

Reply:
I could try to throw in some kind of Stream https://nodejs.org/api/stream.html#stream but that seems roughly as complicated as the helper function I wrote (need to create both `Writable` and `Readable` streams and have them talk to each other)